1. import library yang diperlukan

In [2]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

2. Load dataset

In [3]:
#Read the data
df=pd.read_csv('news.csv')

#Get shape and head
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

3. Dapatkan label dari DataFrame

In [4]:
#DataFlair - Get the labels
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

4. Bagi dataset menjadi data training dan data testing

In [5]:
#DataFlair - Split the dataset
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

5. Mari kita menginisialisasi TfidfVectorizer dengan kata-kata berhenti dari bahasa Inggris dan frekuensi dokumen maksimum 0,7 (istilah dengan frekuensi dokumen yang lebih tinggi akan dibuang). Hentikan kata adalah kata yang paling umum dalam bahasa yang akan disaring sebelum memproses data bahasa alami. Dan TfidfVectorizer mengubah kumpulan dokumen mentah menjadi matriks fitur TF-IDF.

In [6]:
#DataFlair - Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

#DataFlair - Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

6.  Selanjutnya, kita akan menginisialisasi PassiveAggressiveClassifier. Ini. Kami akan memasang ini di tfidf_train dan y_train.

Kemudian, kami akan memprediksi pada set pengujian dari TfidfVectorizer dan menghitung akurasi dengan akurasi_score() dari sklearn.metrics.

In [7]:
#DataFlair - Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

#DataFlair - Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.58%


7. Kami mendapat akurasi 92,82% dengan model ini. Terakhir, mari kita cetak matriks kebingungan untuk mendapatkan wawasan tentang jumlah negatif dan positif palsu dan benar

In [8]:
#DataFlair - Build confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[586,  52],
       [ 42, 587]])

Jadi dengan model ini, kita memiliki 589 positif sejati, 587 negatif sejati, 42 positif palsu, dan 49 negatif palsu.

**Ringkasan**

Hari ini, kami belajar mendeteksi berita palsu dengan Python. Kami mengambil kumpulan data politik, menerapkan TfidfVectorizer, menginisialisasi PassiveAggressiveClassifier, dan menyesuaikan model kami. Kami akhirnya memperoleh akurasi sebesar 92,82%.

Semoga Anda menikmati proyek python deteksi berita palsu. Kunjungi terus DataFlair untuk proyek python, ilmu data, dan pembelajaran mesin yang lebih menarik.

**Apa itu TfidfVectorizer?**

**TF (Frekuensi Jangka)**: Jumlah kali sebuah kata muncul dalam dokumen adalah Frekuensi Jangka-nya. Nilai yang lebih tinggi berarti suatu istilah muncul lebih sering daripada yang lain, sehingga dokumen cocok jika istilah tersebut merupakan bagian dari istilah pencarian.

**IDF (Frekuensi Dokumen Terbalik)**: Kata-kata yang muncul berkali-kali dalam dokumen, tetapi juga muncul berkali-kali di banyak dokumen lainnya, mungkin tidak relevan. IDF adalah ukuran seberapa signifikan suatu istilah di seluruh korpus.

TfidfVectorizer mengubah kumpulan dokumen mentah menjadi matriks fitur TF-IDF.

**PassiveAggressiveClassifier** : Algoritma Passive Aggressive adalah algoritma pembelajaran online. Algoritme seperti itu tetap pasif untuk hasil klasifikasi yang benar, dan berubah menjadi agresif jika terjadi kesalahan perhitungan, pemutakhiran, dan penyesuaian. Tidak seperti kebanyakan algoritma lainnya, ini tidak konvergen. Tujuannya adalah untuk membuat pembaruan yang mengoreksi kerugian, menyebabkan perubahan yang sangat kecil pada norma vektor bobot.